### PR TITLE
mimalloc: enable testing; fix build for older systems and PPC

### PR DESCRIPTION
**Pull request by barracuda156**
#### Description

Fixes: https://trac.macports.org/ticket/67220

###### Type(s)

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.6
Xcode 3.2

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [ ] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

**Comment by barracuda156** [comment]
```
--->  Testing mimalloc
Executing:  cd "/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_mimalloc/mimalloc/work/build" && /usr/bin/make test 
Running tests...
/opt/local/bin/ctest --force-new-ctest-process 
Test project /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_mimalloc/mimalloc/work/build
    Start 1: test-api
1/3 Test #1: test-api .........................   Passed    0.69 sec
    Start 2: test-api-fill
2/3 Test #2: test-api-fill ....................   Passed    0.01 sec
    Start 3: test-stress
3/3 Test #3: test-stress ......................   Passed    0.99 sec

100% tests passed, 0 tests failed out of 3

Total Test time (real) =   1.70 sec
```

File: devel/mimalloc/Portfile [hunk 1]
```diff
@@ -22,14 +26,27 @@ checksums           rmd160  362b6ba6270889af10af46e41e469847a42b99a0 \
...
+                    0001-prim.h-support-tls_slot-for-PPC.patch \
+                    0002-Easy-on-tests-for-PPC.patch
 
 compiler.cxx_standard   2017
 compiler.c_standard     2011
 
 configure.args-append \
                     -DMI_BUILD_TESTS=ON
 
+if {[string match *gcc* ${configure.compiler}]} {
```
Comment (kencu) [inline devel/mimalloc/Portfile:39]: you think gcc being the compiler causes the issue? That would be unexpected…

File: devel/mimalloc/files/0001-prim.h-support-tls_slot-for-PPC.patch [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+diff --git include/mimalloc/prim.h include/mimalloc/prim.h
+index 9e56069..30be1af 100644
+--- include/mimalloc/prim.h
++++ include/mimalloc/prim.h
+@@ -8,6 +8,9 @@ terms of the MIT license. A copy of the license can be found in the file
+ #ifndef MIMALLOC_PRIM_H
+ #define MIMALLOC_PRIM_H
+ 
++#if defined(__APPLE__) && defined(__POWERPC__)
```
Comment (kencu) [inline devel/mimalloc/files/0001-prim.h-support-tls_slot-for-PPC.patch:18]: only PowerPC systems need pthread.h? That would be extremely odd…

Comment (barracuda156) [reply devel/mimalloc/files/0001-prim.h-support-tls_slot-for-PPC.patch:18]: Intel uses assembler there, I would assume it works regardless of SDK or compiler.

File: devel/mimalloc/Portfile (hunk 1 above)
Comment (barracuda156) [reply devel/mimalloc/Portfile:39]: Clang won’t link with `libatomic`, and GCC is only used by default on PPC.
But you are right, perhaps it makes sense to separate args into separate groups with different conditions.

I have no idea though which OS range should be used. Someone needs to run tests and compare what works better on Intel, I guess.

File: devel/mimalloc/files/0001-prim.h-support-tls_slot-for-PPC.patch (hunk 2 above)
Comment (kencu) [reply devel/mimalloc/files/0001-prim.h-support-tls_slot-for-PPC.patch:18]: Have you got any other systems to check these patches on? Given the systems you are interested in, you’d probably want to test them on Leopard i386, standard SnowLeopard Intel, and then run them past a current system.

You often try to make your patches specific to your systems of interest, but the kinds of patches you are doing most often don’t appear to me to be specific to gcc as a compiler or to powerpc as an architecture.

File: devel/mimalloc/Portfile (hunk 1 above)
Comment (kencu) [reply devel/mimalloc/Portfile:39]: Like the song goes…. “That someone…. is you…” :)

No, really though. It’s time to kick these fixes up a notch, really apply them where they are needed, and make sure they are right.

It’s easy to set up a Leopard i386 VM, many of us have one. You already have SnowLeopard going.

File: devel/mimalloc/files/0001-prim.h-support-tls_slot-for-PPC.patch (hunk 2 above)
Comment (barracuda156) [reply devel/mimalloc/files/0001-prim.h-support-tls_slot-for-PPC.patch:18]: I have no environment to test Intel builds, that too something like Leopard i386. I agree with you that here the patch isn’t GCC-specific, at least aside of `libatomic` part, but the choice is between fixing something that is actually tested and blind shot in a hope something else gets fixed along. I am fine with the latter though.

File: devel/mimalloc/Portfile (hunk 1 above)
Comment (barracuda156) [reply devel/mimalloc/Portfile:39]: Well, I believe in benefits of division of labor :)

Likewise, pretty no one is testing anything on PPC when either submitting new ports, updating existing ones or making fixes, and I do not expect everyone installing 10.5.8 / 10.6.8 Server in a VM to do that testing (not a complicated thing to do though, right?), similarly, I am neither interested nor have time and other resources to maintain another dozen of build systems. Nor see a point – there are plenty of members here who are able to test on Intel, and much better than me. For me it is a waste of time – at least in a sense I rather use the same for something that others won’t do.

**Comment by barracuda156** [comment]
Forgot to make tests patch conditional, will make that later.

File: devel/mimalloc/Portfile (hunk 1 above)
Comment (kencu) [reply devel/mimalloc/Portfile]: Well, you've stated your position, at least :) 

Your patches will only ever be tested by you on your custom setups. The rest of MacPorts can just hope they work, or follow along behind you and try to fix them properly if they are broken.

That will, indeed, make them much more difficult for anyone to commit. But at least you're very clear about where you stand.

Comment (barracuda156) [reply devel/mimalloc/Portfile]: From the very fact that no one committed a fix for Intel systems for this port we can deduce that either a) everything works fine on Intel, and my patches can be restricted to PowerPC (and then the problem you pointed to does not exist), or b) no one cares about older Intel systems and no one ever bothered to test the port on those (and then why should we).
In any case, I address a specific bug – the port is currently broken on PPC systems. If there is any concern about applicability of my patches to Intel, I eagerly make everything PPC-only.

**Comment by barracuda156** [comment]
@mohd-akram As the maintainer, would you prefer restricting the changes to PowerPC systems (where those are tested and confirmed to work) or rather applying those to older systems of any arch? I have no specific preference here, I just want to have this unbroken on PowerPC.

File: devel/mimalloc/Portfile (hunk 1 above)
Comment (barracuda156) [reply devel/mimalloc/Portfile]: @kencu On the other hand, if you are actually concerned about the port in question and not just enjoying an argument, why dont’t you run tests for it? And then we find out if Intel needs any fixes or not :)

File: devel/mimalloc/files/0001-prim.h-support-tls_slot-for-PPC.patch (hunk 2 above)
Comment (mohd-akram) [reply devel/mimalloc/files/0001-prim.h-support-tls_slot-for-PPC.patch:18]: There are other references to pthread in that file. pthread.h is included by internal.h, and so it should be picked up by prim.h.

File: devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+-  CHECK_BODY("realpath") {
+-    char* s = mi_realpath( ".", NULL );
+-    // printf("realpath: %s\n",s);
+-    mi_free(s);
+-  };
++//   CHECK_BODY("realpath") {
++//     char* s = mi_realpath( ".", NULL );
++//     // printf("realpath: %s\n",s);
++//     mi_free(s);
++//   };
```
Comment (mohd-akram) [inline devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch:28]: Why is this commented?

File: devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+diff --git test/test-stress.c test/test-stress.c
+index 3ecb67b..6d90d00 100644
+--- test/test-stress.c
++++ test/test-stress.c
+@@ -25,7 +25,7 @@ terms of the MIT license.
+ // > mimalloc-test-stress [THREADS] [SCALE] [ITER]
+ //
+ // argument defaults
+-static int THREADS = 32;      // more repeatable if THREADS <= #processors
++static int THREADS = 4;      // more repeatable if THREADS <= #processors
```
Comment (mohd-akram) [inline devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch:41]: What's the issue here?

File: devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ static int SCALE   = 25;      // scaling factor
+ 
+ #if defined(MI_TSAN)
+--
```
Comment (mohd-akram) [inline devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch:45]: Remove.

**Comment by mohd-akram** [comment]
If the build was broken anyway for those older systems, then it's fine to patch them too unless something is clearly PowerPC specific.

File: devel/mimalloc/files/0001-prim.h-support-tls_slot-for-PPC.patch [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+     __asm__ volatile ("mrs %0, tpidr_el0" : "=r" (tcb));
+     #endif
+     tcb[slot] = value;
++  #elif defined(__POWERPC__)
++    int res; MI_UNUSED(ofs);
++    res = pthread_setspecific(slot, value);
+   #endif
+ }
+ 
+--
```
Comment (mohd-akram) [inline devel/mimalloc/files/0001-prim.h-support-tls_slot-for-PPC.patch:53]: Remove.

File: devel/mimalloc/Portfile (hunk 1 above)
Comment (mohd-akram) [reply devel/mimalloc/Portfile]: I doubt this is strictly GCC related too. I think put that condition underneath the os version check as well.

File: devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch (hunk 4 above)
Comment (barracuda156) [reply devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch:41]: The patch should not be universally applied, I just forgot to make it conditional, will do it.
On PPC machine the tests takes forever – I gave up after an 1.5 hrs+ of waiting.

File: devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch (hunk 3 above)
Comment (barracuda156) [reply devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch:28]: This test case fails on PPC. (Again, should not be applied to other systems, sorry, I was too tired yesterday.)

**Comment by barracuda156** [comment]
@mohd-akram I will make recommended changes and rebase.

File: devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch (hunk 3 above)
Comment (mohd-akram) [reply devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch:28]: No worries. I'm curious why it fails though, it seems like a simple test.

File: devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch (hunk 4 above)
Comment (mohd-akram) [reply devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch:41]: Makes sense.

File: devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch (hunk 3 above)
Comment (barracuda156) [reply devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch:28]: BTW it passed fine when `MI_OSX_ZONE` is on, but then stress test is broken.

**Comment by barracuda156** [comment]
@mohd-akram Please take a look.

File: devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch (hunk 3 above)
Comment (barracuda156) [reply devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch:33]: I added comments to the patch re changes done.

Comment (mohd-akram) [reply devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch:33]: This is strange tbh. This function is part of the public API so it should work. See if you can investigate further - which line in the source does it break on?

Comment (mohd-akram) [reply devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch:33]: The source for the function is [here](https://github.com/microsoft/mimalloc/blob/v2.1.2/src/alloc.c#L864). It doesn't do anything particularly crazy.

**Comment by mohd-akram** [comment]
Still missing some ifdefs around tests, and putting the gcc check inside the os version check. Also, for the stress test, can you try commenting the two lines of THREADS and SCALE and uncommenting the ones below which have a value of 8 and 100 instead?

**Comment by barracuda156** [comment]
@mohd-akram So enabling that realpath test case, I get:
```
--->  Testing mimalloc
Executing:  cd "/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_mimalloc/mimalloc/work/build" && /usr/bin/make test 
Running tests...
/opt/local/bin/ctest --force-new-ctest-process 
Test project /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_mimalloc/mimalloc/work/build
    Start 1: test-api
1/3 Test #1: test-api .........................***Exception: SegFault  1.30 sec
    Start 2: test-api-fill
2/3 Test #2: test-api-fill ....................   Passed    0.01 sec
    Start 3: test-stress
3/3 Test #3: test-stress ......................   Passed    1.45 sec

67% tests passed, 1 tests failed out of 3

Total Test time (real) =   2.77 sec

The following tests FAILED:
	  1 - test-api (SEGFAULT)
Errors while running CTest
```
```
1/3 Testing: test-api
1/3 Test: test-api
Command: "/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_mimalloc/mimalloc/work/build/mimalloc-test-api"
Directory: /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_mimalloc/mimalloc/work/build
"test-api" start time: Jul 08 17:26 CST
Output:
----------------------------------------------------------
test: malloc-zero...  ok.
test: malloc-nomem1...  ok.
test: malloc-null...  ok.
test: calloc-overflow...  ok.
test: calloc0...  ok.
test: malloc-large...  ok.
test: posix_memalign1...  ok.
test: posix_memalign_no_align...  ok.
test: posix_memalign_zero...  ok.
test: posix_memalign_nopow2...  ok.
test: posix_memalign_nomem...  ok.
test: malloc-aligned1...  ok.
test: malloc-aligned2...  ok.
test: malloc-aligned3...  ok.
test: malloc-aligned4...  ok.
test: malloc-aligned5...  ok.
test: malloc-aligned6...  ok.
test: malloc-aligned7...  ok.
test: malloc-aligned8...  ok.
test: malloc-aligned9...  ok.
test: malloc-aligned10...  ok.
test: malloc_aligned11...  ok.
test: mimalloc-aligned12...  ok.
test: malloc-aligned-at1...  ok.
test: malloc-aligned-at2...  ok.
test: memalign1...  ok.
test: zalloc-aligned-small1...  ok.
test: rezalloc_aligned-small1...  ok.
test: realloc-null...  ok.
test: realloc-null-sizezero...  ok.
test: realloc-sizezero...  ok.
test: reallocarray-null-sizezero...  ok.
test: heap_destroy...  ok.
test: heap_delete...  ok.
test: realpath...  
<end of output>
Test time =   1.30 sec
----------------------------------------------------------
Test Failed.
"test-api" end time: Jul 08 17:26 CST
"test-api" time elapsed: 00:00:01
----------------------------------------------------------
```
```
Process:         mimalloc-test-api [26087]
Path:            /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_mimalloc/mimalloc/work/build/mimalloc-test-api
Identifier:      mimalloc-test-api
Version:         ??? (???)
Code Type:       PPC (Native)
Parent Process:  ctest-orig [26086]

Date/Time:       2023-07-08 17:26:09.851 +0800
OS Version:      Mac OS X 10.6 (10A190)
Report Version:  6

Exception Type:  EXC_BAD_ACCESS (SIGSEGV)
Exception Codes: KERN_INVALID_ADDRESS at 0x0000000096d19b68
Crashed Thread:  0

Thread 0 Crashed:
0   libmimalloc.2.dylib           	0x00023b3c _mi_free_generic + 0
1   mimalloc-test-api             	0x00004ed0 main + 3748

Thread 0 crashed with PPC Thread State 32:
  srr0: 0x00023b3c  srr1: 0x0200d030   dar: 0x96d19b68 dsisr: 0x00200000
    r0: 0x0002c720    r1: 0xbffff190    r2: 0x00400600    r3: 0x00400000
    r4: 0x96d19b5c    r5: 0x00000000    r6: 0x00500340    r7: 0x6f706172
    r8: 0x6f632f77    r9: 0x6e672065   r10: 0x0000067c   r11: 0x00032070
   r12: 0x00254310   r13: 0xbffff260   r14: 0x00000000   r15: 0x00000001
   r16: 0xbffff244   r17: 0x00000000   r18: 0x01000000   r19: 0x00400000
   r20: 0x00200001   r21: 0x00200000   r22: 0x00100000   r23: 0x00000200
   r24: 0x00000008   r25: 0x00000001   r26: 0x0107801c   r27: 0x00003f9e
   r28: 0x00003f95   r29: 0x010600c0   r30: 0x00500340   r31: 0x0000403c
    cr: 0x82000259   xer: 0x20000000    lr: 0x0002c720   ctr: 0x00254310
vrsave: 0x00000000

Binary Images:
    0x1000 -     0x4fff +mimalloc-test-api ??? (???) <131d64d1608b1975b1be3e24308969c0> /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_mimalloc/mimalloc/work/build/mimalloc-test-api
    0xa000 -    0x10fff +libMacportsLegacySupport.dylib ??? (???) <ca69634c379c16188f685cc83274e358> /opt/local/lib/libMacportsLegacySupport.dylib
   0x14000 -    0x16fff +libatomic.1.dylib ??? (???) <e92b0dacdfb8a2ffe0a72bb730f56ba2> /opt/local/lib/libgcc/libatomic.1.dylib
   0x21000 -    0x31fdb +libmimalloc.2.dylib ??? (???) <44d7a5c294c80dda9bf1faed14670ee6> /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_mimalloc/mimalloc/work/build/libmimalloc.2.dylib
   0x42000 -    0x42fff +libgcc_s.1.dylib ??? (???) <52f8e6b93cfbb9fa40f44e2651069377> /opt/local/lib/libgcc/libgcc_s.1.dylib
   0x45000 -    0x4aff7 +libgcc_s.1.1.dylib ??? (???) <271e374b41239a540bc3b6cb497bbe03> /opt/local/lib/libgcc/libgcc_s.1.1.dylib
   0x4d000 -    0x52fff +libgcc_ehs.1.1.dylib ??? (???) <3afeacbd27bb3f27ddfc690543cd1b47> /opt/local/lib/libgcc/libgcc_ehs.1.1.dylib
   0x55000 -    0x5aff6  libmathCommon.A.dylib ??? (???) <e166c8e59a80a82b4504b1dd43d1f50a> /usr/lib/system/libmathCommon.A.dylib
  0x252000 -   0x41cfe7  libSystem.B.dylib ??? (???) <420b38fe0fa0944f6f5bc4840bf29a4e> /usr/lib/libSystem.B.dylib
0x8fe00000 - 0x8fe31143  dyld 113.0 (???) <ee555e655e1cf66cac80728801add419> /usr/lib/dyld
0xffff8000 - 0xffff9703  libSystem.B.dylib ??? (???) /usr/lib/libSystem.B.dylib
```

File: devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch (hunk 3 above)
Comment (barracuda156) [reply devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch:33]: Let me try with the debugger.

**Comment by barracuda156** [comment]
@mohd-akram It fails here:
```
Program received signal EXC_BAD_ACCESS, Could not access memory.
Reason: KERN_INVALID_ADDRESS at address: 0x96d19b68
0x00023b3c in _mi_free_generic ()
(gdb) where
#0  0x00023b3c in _mi_free_generic ()
#1  0x0002c720 in mi_heap_realpath () at _string.h:80
#2  0x00004ed4 in main ()
```

**Comment by mohd-akram** [comment]
Running:

```
cmake -DMI_OSX_INTERPOSE=OFF -DMI_OSX_ZONE=OFF -DMI_USE_CXX=OFF ../..
make && make test
```

causes it to fail for me too on an Intel system (Ventura), so it seems like an upstream issue. Looks like it's due to the fact that the `free` inside mi_heap_realpath is redirected to mimalloc and it doesn't know what to do with it (despite the comment there saying it's fine).

**Comment by barracuda156** [comment]
@mohd-akram Thank you very much for checking!

What do you suggest for now (i.e. with the port)? I can open an issue with upstream for the error.

**Comment by mohd-akram** [comment]
Remove the patch for that test, it's fine if it fails. If you'll open an issue, you can put a link to it above the `test.run` line mentioning the expected failure. Also, curious why do you add `-DMI_USE_CXX=OFF`?

FYI, issue happens consistently with just `-DMI_OSX_ZONE=OFF`.

**Comment by mohd-akram** [comment]
Actually, just append the patchfile conditionally next to where you're adding the problematic configure.args.

**Comment by barracuda156** [comment]
@mohd-akram Re why MI_USE_CXX=OFF: I know it is not ON by default anyway, but this is a safe-guard. For w/e reason turning it on breaks linking with LegacySupport:
```
/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_mimalloc/mimalloc/work/mimalloc-2.1.2/src/prim/osx/../unix/prim.c: In function 'bool _mi_prim_random_buf(void*, size_t)':
/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_mimalloc/mimalloc/work/mimalloc-2.1.2/src/prim/osx/../unix/prim.c:746:5: error: 'arc4random_buf' was not declared in this scope
  746 |     arc4random_buf(buf, buf_len);
      |     ^~~~~~~~~~~~~~
```

File: devel/mimalloc/Portfile [hunk 7]
```diff
@@ -23,3 +27,37 @@ checksums           rmd160  362b6ba6270889af10af46e41e469847a42b99a0 \
...
+
+if {[string match *gcc* ${configure.compiler}]} {
+    configure.args-append \
+                    -DMI_USE_LIBATOMIC=ON
+}
+
+# To fix running tests:
+configure.pre_args-replace \
+                    -DCMAKE_BUILD_WITH_INSTALL_RPATH:BOOL=ON \
+                    -DCMAKE_BUILD_WITH_INSTALL_RPATH:BOOL=OFF
```
Comment (mohd-akram) [inline devel/mimalloc/Portfile:61]: Tests run fine for me without this.

File: devel/mimalloc/Portfile [hunk 8]
```diff
@@ -23,3 +27,37 @@ checksums           rmd160  362b6ba6270889af10af46e41e469847a42b99a0 \
...
+    # Turning on MI_OSX_INTERPOSE breaks the build with undefined symbols:
+    # https://trac.macports.org/ticket/67220
+    # Turning on MI_OSX_ZONE breaks test-stress with Bus error.
+    configure.args-append \
+                    -DMI_OSX_INTERPOSE=OFF \
+                    -DMI_OSX_ZONE=OFF \
+                    -DMI_USE_CXX=OFF
+}
+
+if {[string match *gcc* ${configure.compiler}]} {
```
Comment (mohd-akram) [inline devel/mimalloc/Portfile:53]: Move this inside the os version check above.

File: devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+index 3ecb67b..6d90d00 100644
+--- test/test-stress.c
++++ test/test-stress.c
+@@ -25,7 +25,7 @@ terms of the MIT license.
+ // > mimalloc-test-stress [THREADS] [SCALE] [ITER]
+ //
+ // argument defaults
+-static int THREADS = 32;      // more repeatable if THREADS <= #processors
++static int THREADS = 4;      // more repeatable if THREADS <= #processors
+ static int SCALE   = 25;      // scaling factor
```
Comment (mohd-akram) [inline devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch:47]: Try commenting these two lines and uncommenting the ones below with lower THREADS and higher SCALE. It should still be a stress test.

File: devel/mimalloc/Portfile [hunk 10]
```diff
@@ -23,3 +27,37 @@ checksums           rmd160  362b6ba6270889af10af46e41e469847a42b99a0 \
                     size    1163868
 
 patchfiles          patch-cmake.diff
+
+platform darwin powerpc {
+    patchfiles-append \
```
Comment (mohd-akram) [inline devel/mimalloc/Portfile:32]: Move inside os version check below.

File: devel/mimalloc/Portfile [hunk 11]
```diff
@@ -23,3 +27,37 @@ checksums           rmd160  362b6ba6270889af10af46e41e469847a42b99a0 \
...
+                    -DMI_BUILD_TESTS=ON
+
+if {${os.platform} eq "darwin" && ${os.major} < 11} {
+    # Turning on MI_OSX_INTERPOSE breaks the build with undefined symbols:
+    # https://trac.macports.org/ticket/67220
+    # Turning on MI_OSX_ZONE breaks test-stress with Bus error.
+    configure.args-append \
+                    -DMI_OSX_INTERPOSE=OFF \
+                    -DMI_OSX_ZONE=OFF \
+                    -DMI_USE_CXX=OFF
```
Comment (mohd-akram) [inline devel/mimalloc/Portfile:50]: If it's not required, remove it.

**Review by mohd-akram — CHANGES_REQUESTED** [review]

File: devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch (hunk 9 above)
Comment (barracuda156) [reply devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch:47]: Let me try.

File: devel/mimalloc/Portfile (hunk 7 above)
Comment (barracuda156) [reply devel/mimalloc/Portfile:61]: Hmm. I got a failure earlier (and still have related crash reports in front of me), but now I removed it, and tests still ran. No idea why the difference, but let’s remove it, at least until someone complains.

File: devel/mimalloc/Portfile (hunk 11 above)
Comment (barracuda156) [reply devel/mimalloc/Portfile:50]: `-DMI_OSX_INTERPOSE=OFF -DMI_OSX_ZONE=OFF` – these two are required, they are not default, and without them build fails (as noted in comments above).

File: devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch (hunk 9 above)
Comment (barracuda156) [reply devel/mimalloc/files/0002-Easy-on-tests-for-PPC.patch:47]: Yes, that works. I will use that in the tests patch then.

**Comment by barracuda156** [comment]
@mohd-akram Updated according to your suggestions, rebased. Added links to issues with upstream.

**Review by mohd-akram — APPROVED** [review]
LGTM. Thank you!

**Comment by barracuda156** [comment]
> LGTM. Thank you!

Great

**Comment by barracuda156** [comment]
@mohd-akram If we are good on this, could we merge?

**Comment by mohd-akram** [comment]
I would if I could :) Maybe @mascguy can.

**Review by mascguy — APPROVED** [review]
Tested on 10.6 thru 10.9, and builds now succeed. Merging.

**Comment by barracuda156** [comment]
Great, thanks!

**Pull request merged**